Dilated Convolution Layer

yasunorikudo opened the pull request:
I created the issue in #1283. This PR will add dilated convolution layer (https://arxiv.org/abs/1511.07122).

mattya commented:
Hello, thank you for the cool PR.
I've tested your code and it looks working without critical problems.
I have  some comments at this time.

1. It could be helpful if there's more detail description about the relationships between `ksize`, `dilate`, `patch size` and `output size` in the document for `chainer.functions.dilated_convolution_2d`.  
2. It's needed to add DilatedConvolution2D to `chainer/docs/source/reference/links.rst`.
3. The author of the paper says "random initialization schemes were not effective".  Do you have any plans to implement the "identity initialization" proposed in the paper?
4. I briefly benchmarked your function by forward and backward (32, 32, 64, 64) tensor to DilatedConvolution2D(32, 32, ksize=3, stride=1, pad=4, dilate=4, use_cudnn=use_cudnn) and got the result that `use_cudnn=False` was about 2 times faster than `use_cudnn=True`.  Is it likely behaviour?

yasunorikudo commented:
Thank you for your reviewing. I fixed documents and added DilatedConvolution2D to ```chainer/docs/source/reference/links.rst```.
- About identity initialization
  Chainer doesn't support 4D identity Initialization. I will raise an issue about 4D identity initialization and make PR.
- About difference between ```use_cudnn=True``` and ```use_cudnn=False```
  cuDNN doesn't support dilated convolution yet, but I implemented this layer using 1×1 convolution operation in cuDNN library. Therefore in some cases  ```use_cudnn=True``` option is much slower. In my experiments under the same condition above, forward prop is about 50% faster and backward prop is 20% slower than ```use_cudnn=False```.

mattya commented:
Thank you!  LGTM
